src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.bridge.mqtt.core;
+
+import io.strimzi.kafka.bridge.mqtt.utils.MappingRule;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Comparator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Responsible for handling all the topic mapping.
+ *
+ * @see MappingRule
+ */
+public class MqttKafkaMapper {
+
+    // default kafka topic. Used when no mapping rule matches the mqtt topic.
+    public static final String DEFAULT_KAFKA_TOPIC = "messages_default";
+
+    // find any word inside a curly bracket. E.g. {something}, this is known as a placeholder.
+    private static final String MQTT_TOPIC_PLACEHOLDER_REGEX = "\\{\\w+\\}";
+
+    // Regex used to replace the placeholders with {something} in the mqtt pattern.
+    private static final String PLACEHOLDER_REGEX = "(.*)";
+
+    // identifies a single level wildcard character in the mqtt pattern. E.g. sensors/+/data
+    private static final String MQTT_TOPIC_SINGLE_LEVEL_WILDCARD_CHARACTER = "+";
+
+    // Regex expression used to replace the + in the mqtt pattern.
+    private static final String SINGLE_LEVEL_WILDCARD_REGEX = "[^/]+";
+
+    // identifies a multi level wildcard character in the mqtt pattern. E.g. sensors/#
+    private static final String MQTT_TOPIC_MULTI_LEVEL_WILDCARD_CHARACTER = "#";
+
+    // matches any character after the string. Used to replace the # in the mqtt pattern.
+    private static final String MULTIPLE_LEVEL_WILDCARD_REGEX = ".*";
+    private final List<MappingRule> rules;
+    private final List<Pattern> patterns = new ArrayList<>();
+    private Pattern placeholderPattern;
+
+    /**
+     * Constructor
+     * <p>
+     * Creates a new instance of MqttKafkaMapper.
+     */
+    public MqttKafkaMapper(List<MappingRule> rules) {
+        this.rules = rules;
+        this.rules.sort(Comparator.comparing(MappingRule::getMqttTopicPatternLevels).reversed());
+        buildRegex();
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            Matcher matcher = this.patterns.get(this.rules.indexOf(rule)).matcher(mqttTopic);
+
+            if (matcher.matches()) {
+                HashMap<String, String> placeholders = new HashMap<>();
+                String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
```

Comment:
this could be improved ... the idea is, using the placeholder pattern to match only the placeholders in the MQTT pattern instead of bringing with you pieces which are not placeholder but you still check in the next two loops.
For example, having `/buildings/{building}/rooms/{room}`, if you just split you will get the array will all 4 parts `[buildings, {building}, rooms, {room}]` while what you actually needs is just `[{building}, {room}]` with just placeholders. This way you don't have to loop across useless element in the array in the next two loops.

Reply:
makes sense!

Reply:
I was just doing it.